cylc/flow/dbstatecheck.py

```diff
@@ -138,49 +255,103 @@ def workflow_state_query(
             stmt_args.append(cycle)
 
-        if status:
-            stmt_frags = []
-            for state in self.state_lookup(status):
-                stmt_args.append(state)
-                stmt_frags.append("status==?")
-            stmt_wheres.append("(" + (" OR ").join(stmt_frags) + ")")
+        if (
+            selector is not None
+            and target_table == CylcWorkflowDAO.TABLE_TASK_STATES
+        ):
+            # Can select by status in the DB but not outputs.
+            stmt_wheres.append("status==?")
+            stmt_args.append(selector)
+
         if stmt_wheres:
-            stmt += " where " + (" AND ").join(stmt_wheres)
+            stmt += "WHERE\n    " + (" AND ").join(stmt_wheres)
+
+        if target_table == CylcWorkflowDAO.TABLE_TASK_STATES:
+            # (outputs table doesn't record submit number)
+            stmt += r"ORDER BY submit_num"
 
-        res = []
+        # Query the DB and drop incompatible rows.
+        db_res = []
         for row in self.conn.execute(stmt, stmt_args):
-            if not all(v is None for v in row):
-                res.append(list(row))
-
-        return res
-
-    def task_state_getter(self, task, cycle):
-        """used to get the state of a particular task at a particular cycle"""
-        return self.workflow_state_query(task, cycle, mask="status")[0]
-
-    def task_state_met(self, task, cycle, status=None, message=None):
-        """used to check if a task is in a particular state"""
-        res = self.workflow_state_query(task, cycle, status, message)
-        if status:
-            return bool(res)
-        elif message:
-            return any(
-                message == value
-                for outputs_str, in res
-                for value in json.loads(outputs_str)
-            )
+            # name, cycle, status_or_outputs, [flow_nums]
+            res = list(row[:3])
+            if row[2] is None:
+                # status can be None in Cylc 7 DBs
+                continue
+            if not self.c7_back_compat_mode:
+                flow_nums = deserialise_set(row[3])
+                if flow_num is not None and flow_num not in flow_nums:
+                    # skip result, wrong flow
+                    continue
+                fstr = stringify_flow_nums(flow_nums)
+                if fstr:
+                    res.append(fstr)
+            db_res.append(res)
+
+        if target_table == CylcWorkflowDAO.TABLE_TASK_STATES:
+            return db_res
+
+        warn_output_fallback = is_trigger
+        results = []
+        for row in db_res:
+            outputs: Union[Dict[str, str], List[str]] = json.loads(row[2])
+            if isinstance(outputs, dict):
+                messages: Iterable[str] = outputs.values()
+            else:
+                # Cylc 8 pre 8.3.0 back-compat: list of output messages
+                messages = outputs
+                if warn_output_fallback:
+                    LOG.warning(output_fallback_msg)
```

Comment:
This message doesn't get through when using the `workflow_state` xtrigger. It appears the logging handler is the null handler in this case, not sure why, here is a workaround
```suggestion
                    print(f"WARNING - {output_fallback_msg}", file=sys.stderr)
```

Reply:
I've pushed it